### PR TITLE
octopus: mgr/dashboard: additional logging to SMART data retrieval

### PR DESCRIPTION
**Pull request by callithea**
backport trackers:

* https://tracker.ceph.com/issues/48131
* https://tracker.ceph.com/issues/48500

---

backport of https://github.com/ceph/ceph/pull/37637
parent tracker: https://tracker.ceph.com/issues/47834

this backport was staged using ceph-backport.sh version 15.1.1.389
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

File: src/pybind/mgr/dashboard/services/ceph_service.py
Comment (bk201) [inline src/pybind/mgr/dashboard/services/ceph_service.py:306]:
```suggestion
                'with ID %s'
```

Maybe we can fix this issue in the PR without creating a backport: https://tracker.ceph.com/issues/48245

Comment (callithea) [reply src/pybind/mgr/dashboard/services/ceph_service.py:306]: We could also wait until the PR targeting master has been merged and I could cherry-pick the commit into this PR, what do you think?

Comment (bk201) [reply src/pybind/mgr/dashboard/services/ceph_service.py:306]: That will make perfect sense to me! Thanks for waiting.

**Comment by callithea** [comment]
Waiting for #38163 to get merged so the commit can be cherry-picked into this backport PR.

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test dashboard

**Comment by smithfarm** [comment]
@callithea I guess DNM can be dropped now?

**Comment by callithea** [comment]
> @callithea I guess DNM can be dropped now?

Unfortunately `make check` is still failing, I'm looking into that and will remove the DNM label when it's fixed.

**Comment by callithea** [comment]
`make check` failure:
```
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
src/pybind/mgr/dashboard/CMakeFiles/mgr-dashboard-venv.dir/build.make:60: recipe for target 'mgr-dashboard-virtualenv/bin/activate' failed
make[3]: *** [mgr-dashboard-virtualenv/bin/activate] Error 1
make[3]: *** Deleting file 'mgr-dashboard-virtualenv/bin/activate'
CMakeFiles/Makefile2:5858: recipe for target 'src/pybind/mgr/dashboard/CMakeFiles/mgr-dashboard-venv.dir/all' failed
make[2]: *** [src/pybind/mgr/dashboard/CMakeFiles/mgr-dashboard-venv.dir/all] Error 2
```

**Comment by smithfarm** [comment]
Hm, somewhat further up in the make check log I found this:

```
    ERROR: Command errored out with exit status 1:
     command: /home/jenkins-build/build/workspace/ceph-pull-requests/install-deps-python3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-wheel-fhec2_13/pyfakefs_26afa90f64c94b5c94175e63dba3ff5d/setup.py'"'"'; __file__='"'"'/tmp/pip-wheel-fhec2_13/pyfakefs_26afa90f64c94b5c94175e63dba3ff5d/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-8wu7iojv
         cwd: /tmp/pip-wheel-fhec2_13/pyfakefs_26afa90f64c94b5c94175e63dba3ff5d/
    Complete output (8 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-wheel-fhec2_13/pyfakefs_26afa90f64c94b5c94175e63dba3ff5d/setup.py", line 3, in <module>
        from fake_filesystem import __version__
      File "/tmp/pip-wheel-fhec2_13/pyfakefs_26afa90f64c94b5c94175e63dba3ff5d/fake_filesystem.py", line 100
        PERM_READ = 0400      # Read permission bit.
                       ^
    SyntaxError: invalid token
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
@djgalloway @tchaikov Do you know what to "make" of the "make check" log in this PR? It starts with this:

```
+ timeout 3h ./run-make-check.sh
Checking hostname sanity... OK
Reading package lists...
Building dependency tree...
Reading state information...
ccache is already the newest version (3.4.1-1).
debianutils is already the newest version (4.8.4).
The following packages were automatically installed and are no longer required:
  ceph-libboost-atomic1.72.0 ceph-libboost-chrono1.72.0
  ceph-libboost-container1.72.0 ceph-libboost-context1.72.0
  ceph-libboost-coroutine1.72.0 ceph-libboost-date-time1.72.0
  ceph-libboost-filesystem1.72.0 ceph-libboost-iostreams1.72.0
  ceph-libboost-program-options1.72.0 ceph-libboost-python1.72.0
  ceph-libboost-random1.72.0 ceph-libboost-regex1.72.0
  ceph-libboost-serialization1.72.0 ceph-libboost-system1.72.0
  ceph-libboost-test1.72.0 ceph-libboost-thread1.72.0
  ceph-libboost-timer1.72.0 cryptsetup-bin cython dh-exec dh-systemd
...
```

which doesn't look right. And further down we see: 

```
Suggested packages:
  ceph-libboost1.72-doc gccxml ceph-libboost-exception1.72-dev
  ceph-libboost-fiber1.72-dev ceph-libboost-graph1.72-dev
  ceph-libboost-graph-parallel1.72-dev ceph-libboost-locale1.72-dev
  ceph-libboost-log1.72-dev ceph-libboost-math1.72-dev
  ceph-libboost-stacktrace1.72-dev ceph-libboost-type-erasure1.72-dev
  ceph-libboost-wave1.72-dev ceph-libboost1.72-tools-dev libmpfrc++-dev
  libntl-dev
The following NEW packages will be installed:
  ceph-libboost-atomic1.72-dev ceph-libboost-chrono1.72-dev
  ceph-libboost-container1.72-dev ceph-libboost-context1.72-dev
  ceph-libboost-coroutine1.72-dev ceph-libboost-date-time1.72-dev
  ceph-libboost-filesystem1.72-dev ceph-libboost-iostreams1.72-dev
  ceph-libboost-program-options1.72-dev ceph-libboost-python1.72-dev
  ceph-libboost-random1.72-dev ceph-libboost-regex1.72-dev
  ceph-libboost-serialization1.72-dev ceph-libboost-system1.72-dev
  ceph-libboost-test1.72-dev ceph-libboost-thread1.72-dev
  ceph-libboost-timer1.72-dev ceph-libboost1.72-dev
0 upgraded, 18 newly installed, 0 to remove and 240 not upgraded.
Need to get 0 B/14.4 MB of archives.
After this operation, 190 MB of additional disk space will be used.
(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 147779 files and directories currently installed.)
Preparing to unpack .../00-ceph-libboost1.72-dev_1.72.0-1_amd64.deb ...
Unpacking ceph-libboost1.72-dev:amd64 (1.72.0-1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-do21W2/00-ceph-libboost1.72-dev_1.72.0-1_amd64.deb (--unpack):
 trying to overwrite '/opt/ceph/include/boost/accumulators/accumulators.hpp', which is also in package ceph-libboost1.73-dev:amd64 1.73.0-1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Preparing to unpack .../01-ceph-libboost-atomic1.72-dev_1.72.0-1_amd64.deb ...
Unpacking ceph-libboost-atomic1.72-dev:amd64 (1.72.0-1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-do21W2/01-ceph-libboost-atomic1.72-dev_1.72.0-1_amd64.deb (--unpack):
 trying to overwrite '/opt/ceph/lib/x86_64-linux-gnu/libboost_atomic.a', which is also in package ceph-libboost-atomic1.73-dev:amd64 1.73.0-1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Preparing to unpack .../02-ceph-libboost-chrono1.72-dev_1.72.0-1_amd64.deb ...
Unpacking ceph-libboost-chrono1.72-dev:amd64 (1.72.0-1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-do21W2/02-ceph-libboost-chrono1.72-dev_1.72.0-1_amd64.deb (--unpack):
 trying to overwrite '/opt/ceph/lib/x86_64-linux-gnu/libboost_chrono.a', which is also in package ceph-libboost-chrono1.73-dev:amd64 1.73.0-1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Preparing to unpack .../03-ceph-libboost-container1.72-dev_1.72.0-1_amd64.deb ...
Unpacking ceph-libboost-container1.72-dev:amd64 (1.72.0-1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-do21W2/03-ceph-libboost-container1.72-dev_1.72.0-1_amd64.deb (--unpack):
 trying to overwrite '/opt/ceph/lib/x86_64-linux-gnu/libboost_container.a', which is also in package ceph-libboost-container1.73-dev:amd64 1.73.0-1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Preparing to unpack .../04-ceph-libboost-serialization1.72-dev_1.72.0-1_amd64.deb ...
Unpacking ceph-libboost-serialization1.72-dev:amd64 (1.72.0-1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-do21W2/04-ceph-libboost-serialization1.72-dev_1.72.0-1_amd64.deb (--unpack):
 trying to overwrite '/opt/ceph/lib/x86_64-linux-gnu/libboost_serialization.a', which is also in package ceph-libboost-serialization1.73-dev:amd64 1.73.0-1
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
...
```

which also doesn't look right. And finally it fails due to what appears to be a Python syntax error in a module that is pulled into the mgr/dashboard virtualenv as a dependency of the Dashboard's linter (?)

```
recipe for target 'src/pybind/mgr/dashboard/CMakeFiles/mgr-dashboard-venv.dir/all' failed
```

**Comment by djgalloway** [comment]
> @djgalloway @tchaikov Do you know what to "make" of the "make check" log in this PR? It starts with this:

See https://tracker.ceph.com/issues/46208#note-6.  1.72 and 1.73 are conflicting.  1.73 is installed but `install-deps.sh` is insisting on installing 1.72.

None of that is fatal though.  To be honest, I don't know what boost does or if it is at all involved with the dashboard.  For the sake of finding out, I manually removed **all** ceph-libboost packages from the irvingi and re-ran `make check`.  Failed even sooner...?!  https://jenkins.ceph.com/job/ceph-pull-requests/65377/console

```
[ 34%] Built target mds
CMakeFiles/Makefile2:458: recipe for target 'CMakeFiles/tests.dir/rule' failed
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
Makefile:240: recipe for target 'tests' failed
make: *** [tests] Error 2
```

Maybe the baremetal builders are borked.  Trying on an ephemeral builder.  https://jenkins.ceph.com/job/ceph-pull-requests/65378/console

Nope.

```
[100%] Built target rocksdb
[ 21%] Performing install step for 'rocksdb_ext'
[ 21%] Completed 'rocksdb_ext'
[ 21%] Built target rocksdb_ext
CMakeFiles/Makefile2:458: recipe for target 'CMakeFiles/tests.dir/rule' failed
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
Makefile:240: recipe for target 'tests' failed
make: *** [tests] Error 2
```

I need to leave for the day.  Hopefully Kefu can pick this up from here.

**Comment by callithea** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
```
ERROR: Cannot install -r requirements-lint.txt (line 1) and isort==5.5.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested isort==5.5.3
    pylint 2.3.1 depends on isort<5 and >=4.2.5

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```

File: src/pybind/mgr/dashboard/requirements-lint.txt
Comment (smithfarm) [inline src/pybind/mgr/dashboard/requirements-lint.txt:12]: This change breaks `make check` in octopus:

```
ERROR: Cannot install -r requirements-lint.txt (line 1) and isort==5.5.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested isort==5.5.3
    pylint 2.3.1 depends on isort<5 and >=4.2.5

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```

**Review by smithfarm — DISMISSED** [review]

File: src/pybind/mgr/dashboard/requirements-lint.txt
Comment (callithea) [reply src/pybind/mgr/dashboard/requirements-lint.txt:12]: Thanks for your help, @smithfarm!
@bk201 do you think we should only fix `ceph_service.py` instead of cherry-picking the whole commit or should we try it with a different version of isort?

Comment (smithfarm) [reply src/pybind/mgr/dashboard/requirements-lint.txt:12]: I also tend to think the linter bits of this changeset could be dropped, since they would presumably not be needed to fix the issue at hand.

Comment (bk201) [reply src/pybind/mgr/dashboard/requirements-lint.txt:12]: > Thanks for your help, @smithfarm!
> @bk201 do you think we should only fix `ceph_service.py` ([#37972 (comment)](https://github.com/ceph/ceph/pull/37972#discussion_r524046941)) instead of cherry-picking the whole commit or should we try it with a different version of isort?

I think it's OK to ignore the isort.

**Comment by callithea** [comment]
jenkins test api

**Comment by callithea** [comment]
jenkins test docs

**Comment by bk201** [comment]
I see this error when running `install-deps.sh` (not sure why CI passed):

```
ERROR: Double requirement given: pytest<4 (from -r requirements-test.txt (line 2)) (already in pytest (from -r requirements-lint.txt (line 12)), name='pytest')
```

Maybe we need to backport https://github.com/ceph/ceph/commit/128778f25eb64cd334e062d627abdb23e0ef0e49

**Comment by callithea** [comment]
> I see this error when running `install-deps.sh` (not sure why CI passed):
> 
> ```
> ERROR: Double requirement given: pytest<4 (from -r requirements-test.txt (line 2)) (already in pytest (from -r requirements-lint.txt (line 12)), name='pytest')
> ```
> 
> Maybe we need to backport [128778f](https://github.com/ceph/ceph/commit/128778f25eb64cd334e062d627abdb23e0ef0e49)



> I see this error when running `install-deps.sh` (not sure why CI passed):
> 
> ```
> ERROR: Double requirement given: pytest<4 (from -r requirements-test.txt (line 2)) (already in pytest (from -r requirements-lint.txt (line 12)), name='pytest')
> ```
> 
> Maybe we need to backport [128778f](https://github.com/ceph/ceph/commit/128778f25eb64cd334e062d627abdb23e0ef0e49)

@bk201I added the commit to this backport. Do you have time to give it another try?

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test make check

**Comment by bk201** [comment]
> > I see this error when running `install-deps.sh` (not sure why CI passed):
> > ```
> > ERROR: Double requirement given: pytest<4 (from -r requirements-test.txt (line 2)) (already in pytest (from -r requirements-lint.txt (line 12)), name='pytest')
> > ```
> > 
> > 
> > Maybe we need to backport [128778f](https://github.com/ceph/ceph/commit/128778f25eb64cd334e062d627abdb23e0ef0e49)
> 
> @bk201 I added the commit to this backport. Do you have time to give it another try?

After moving to newer `pytest`, some tests break and Kefu did fix that in [commit](https://github.com/ceph/ceph/pull/37692/commits/71979e9b46e21dc3d3cfc6f06f4a84c9b4c7ce78) of the [PR](https://github.com/ceph/ceph/pull/37692/commits). When I try to backport that PR back I found there are some new error and it means there might be some other commits need to be backported too.

One fix might stick to `pytest<4` both in `requirements-lint.txt` and `requirements-test.txt`.

**Review by p-se — APPROVED** [review]
lgtm

**Pull request merged**
